habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -469,6 +469,178 @@ def step(self, *args, is_last_action, **kwargs):
+        else:
+            return spaces.Dict(
+                {
+                    self._action_arg_prefix
+                    + "base_vel": spaces.Box(
+                        shape=(2,), low=-lim, high=lim, dtype=np.float32
+                    )
+                }
+            )
+
+    def collision_check(
+        self, trans, target_trans, center_end_pos, target_rigid_state
+    ):
+        """
+        trans: the transformation of the current location of the robot
+        target_trans: the transformation of the target location of the robot given the center original Navmesh
+        center_end_pos: the original center end position of the robot
+        target_rigid_state: the target state of the robot given the center original Navmesh
+        """
+        num_check_cylinder = len(self._x_offset)
+        # Get the offset positions
+        cur_pos = []
+        goal_pos = []
+        for i in range(num_check_cylinder):
+            cur_pos.append(
+                trans.transform_point(
+                    np.array([self._x_offset[i], 0, self._y_offset[i]])
+                )
+            )
+            goal_pos.append(
+                target_trans.transform_point(
+                    np.array([self._x_offset[i], 0, self._y_offset[i]])
+                )
+            )
+
+        # For step filter of offset positions
+        end_pos = []
+        for i in range(num_check_cylinder):
+            pos = self._sim.step_filter(cur_pos[i], goal_pos[i])
+            # Sanitize the height
+            pos[1] = cur_pos[i][1]
+            end_pos.append(pos)
+
+        # For step filter of the center original pos
+        center_end_pos[1] = trans.translation[1]
+
+        # Planar move distance clamped by NavMesh
+        move = []
+        for i in range(num_check_cylinder):
+            move.append((end_pos[i] - goal_pos[i]).length())
+        center_move = (center_end_pos - trans.translation).length()
+
+        # Wrap the move direction if we use sliding
+        # Find the largest diff moving direction, which means that there is a collision in that cylinder
+        max_idx = np.argmax(move + [center_move])
+        move_vec = (end_pos + [center_end_pos])[max_idx] - (
+            cur_pos + [trans.translation]
+        )[max_idx]
+        new_end_pos = trans.translation + move_vec
+        new_target_trans = mn.Matrix4.from_(
+            target_rigid_state.rotation.to_matrix(), new_end_pos
+        )
+
+        # For detection of linear or angualr velocities
+        # There is a collision if the difference between the clamped NavMesh position and target position is too great for any point.
+        lin_diff = len([v for v in move if v > self._lin_collision_threshold])
+        ang_diff = len([v for v in move if v > self._ang_collision_threshold])
+
+        if (lin_diff > 0) and center_move != 0:
+            return True, new_target_trans
+        elif (ang_diff > 0) and center_move == 0:
+            return True, new_target_trans
```

Comment:
I'm questioning this separation between linear and angular error. It seems to me like you don't need any knowledge of the center position or this distinction. Unless we empirically see a need for different error values, I suggest simplifying and unifying these.

Reply:
Good catch! Thank you, Alex! I unified them.